src/hardware/driver/relay.rs

```diff
@@ -0,0 +1,256 @@
+        let mut mcp = get_mcp(mutex);
+        // set GPIOs to default position
+        mcp.set_gpio(k1_en.into(), Level::Low).unwrap();
+        mcp.set_gpio(k1_en_n.into(), Level::High).unwrap();
+        mcp.set_gpio(k0_d.into(), Level::Low).unwrap();
+        mcp.set_gpio(k0_cp.into(), Level::Low).unwrap();
+
+        let delay = asm_delay::AsmDelay::new(asm_delay::bitrate::Hertz(
+            ccdr.c_ck().to_Hz(),
+        ));
+        Relay {
+            mutex,
+            delay,
+            k1_en_n,
+            k1_en,
+            k0_d,
+            k0_cp,
+        }
+    }
+}
+
+pub mod sm {
+    use super::*;
+    statemachine! {
+        transitions: {
+            *Disabled + Enable / engage_k0 = EnableWaitK0,
+            EnableWaitK0 + RelayDone / disengage_k1 = EnableWaitK1,
+            EnableWaitK1 + RelayDone = Enabled,
+            Enabled + Disable / engage_k1 = DisableWaitK1,
+            DisableWaitK1 + RelayDone / disengage_k0 = DisableWaitK0,
+            DisableWaitK0 + RelayDone = Disabled
+        }
+    }
+}
+
+impl<'a, I2C, E> sm::StateMachineContext for Relay<'a, I2C>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    // set K0 to upper position (note that "upper" and "lower" refer to the schematic)
+    fn engage_k0(&mut self) {
+        let mut mcp = get_mcp(self.mutex);
+        // set flipflop data pin
+        mcp.set_gpio(self.k0_d.into(), Level::High).unwrap();
+        // set flipflop clock input low to prepare rising edge
+        mcp.set_gpio(self.k0_cp.into(), Level::Low).unwrap();
+        self.delay.delay_us(100u32);
+        // set flipflop clock input high to generate rising edge
+        mcp.set_gpio(self.k0_cp.into(), Level::High).unwrap();
+    }
+
+    // set K0 to lower position
+    fn disengage_k0(&mut self) {
+        let mut mcp = get_mcp(self.mutex);
+        mcp.set_gpio(self.k0_d.into(), Level::High).unwrap();
+        mcp.set_gpio(self.k0_cp.into(), Level::Low).unwrap();
+        self.delay.delay_us(100u32);
+        mcp.set_gpio(self.k0_cp.into(), Level::High).unwrap();
+    }
+
+    // set K1 to upper position
+    fn disengage_k1(&mut self) {
+        let mut mcp = get_mcp(self.mutex);
+        // set en high and en _n low in order to engage K1
+        mcp.set_gpio(self.k1_en.into(), Level::Low).unwrap();
+        mcp.set_gpio(self.k1_en_n.into(), Level::High).unwrap();
+    }
+
+    // set K1 to lower position
+    fn engage_k1(&mut self) {
+        let mut mcp = get_mcp(self.mutex);
+        mcp.set_gpio(self.k1_en.into(), Level::High).unwrap();
+        mcp.set_gpio(self.k1_en_n.into(), Level::Low).unwrap();
+    }
+}
+
+impl<'a, I2C, E> sm::StateMachine<Relay<'a, I2C>>
+where
+    I2C: WriteRead<Error = E> + Write<Error = E>,
+    E: Debug,
+{
+    /// Start relay enabling sequence. Returns the relay delay we need to wait for.
+    pub fn enable(&mut self) -> fugit::MillisDuration<u64> {
+        self.process_event(sm::Events::Enable).unwrap();
```

Comment:
Note that this will panic if you call `enable()` or `disable()` (triggered by the user through MQTT at any time) while in one of the `*Wait*` states or in the wrong one of `Disable/Enable`. You should handle the first cases in code and the last one in the state machine (identity transitions `Disabled + Disable = Disabled` etc).

In general go through your code and review every unwrap for possible cases where the user might trigger them.